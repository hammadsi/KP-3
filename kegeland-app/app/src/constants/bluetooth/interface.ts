import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
