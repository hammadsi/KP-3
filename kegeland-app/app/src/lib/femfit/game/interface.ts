import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
