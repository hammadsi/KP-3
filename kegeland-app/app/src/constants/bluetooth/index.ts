import FEMFIT_PROFILE from '~lib/femfit/bluetooth/profile';

import {BluetoothProfile} from './interface';

/**
 * How long to scan for bluetooth peripherals, in seconds.
 */
export const SCAN_TIME = 5;
/**
 * Whether or not to allows duplicate devices
 */
export const ALLOW_DUPLICATE_DEVICES = false;
/**
 * How often to read BLE characteristic notifications
 */
export const UPDATE_INTERVAL_MS = 250;

export const _profiles = {
  femfit: FEMFIT_PROFILE,
} as const;

export type DeviceType = keyof typeof _profiles;

/**
 * Record of supported bluetooth profiles
 */
export const BLE_PROFILES = _profiles as Record<DeviceType, BluetoothProfile>;

/**
 * Record of supported peripherals
 */
export const PERIPHERAL_MAP: Record<string, DeviceType> = {
  [FEMFIT_PROFILE.batteryService.service.uuid]: 'femfit',
};
