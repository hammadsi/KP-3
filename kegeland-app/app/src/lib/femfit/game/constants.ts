import {Dimensions} from 'react-native';

/**
 * Constants for the femfit game
 */
export default {
  MAX_WIDTH: Dimensions.get('screen').width,
  MAX_HEIGHT: Dimensions.get('screen').height,
  BASELINE: Dimensions.get('screen').height * 0.6,
  COINS_PER_SEC: 1,
  HEADER_HEIGHT: 50,
  BOUNDARY_HEIGHT: 50,
  SPRITE_SIZE: 128,
  PLAYER_SIZE: 50,
  COIN_SIZE: 25,
  GRAVITY: 0.4,
};
