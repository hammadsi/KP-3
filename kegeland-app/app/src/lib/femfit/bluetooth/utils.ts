import {clamp, mean, round} from 'lodash';

import {PRESSURE_SENSOR_MAX_VAL, PRESSURE_SENSOR_MIN_VAL} from './constants';

const range = PRESSURE_SENSOR_MAX_VAL - PRESSURE_SENSOR_MIN_VAL;

/**
 * Calculates the pressure increase relative to minimum value
 * @param value pressure data
 * @returns pressure percent
 */
export const pressurePercent = (value: number) =>
  (value - PRESSURE_SENSOR_MIN_VAL) / range;

/**
 * Converts byte data to a battery level
 * @param bytes byte data
 * @returns
 */
export const readBattery = (bytes: number[]) => {
  const voltage =
    // eslint-disable-next-line no-bitwise
    (((((bytes[4] & 255) << 8) | (bytes[5] & 255)) * 1.2) / 1024 / 220000) *
    1220000;

  return clamp(round(voltage / 6.1, 2), 0, 1);
};

/**
 * Converts bytes to a pressure level
 * @param byte1 first byte
 * @param byte2 second byte
 * @returns pressure level
 */
const convertPressure = (byte1: number, byte2: number) => {
  return clamp(
    // eslint-disable-next-line no-bitwise
    ((((byte1 & 255) << 8) | (byte2 & 255)) + 65536) * 0.00750063755,
    PRESSURE_SENSOR_MIN_VAL,
    PRESSURE_SENSOR_MAX_VAL,
  );
};

/**
 * Converts byte to temperature
 * @param byte temperature byte
 * @returns temperature
 */
const convertSensorTemperature = (byte: number) => {
  return byte * 0.16 + 10.0;
};

/**
 * Converts byte data from the femfit sensor to pressures and temperatures for each of the 8 sensors
 * @param data byte data
 * @returns list of pressures and list of temperatures
 */
export const readSensorBytes = (data: number[][]) => {
  const pressures: number[] = [];
  const temps: number[] = [];
  for (const bytes of data) {
    pressures.push(
      ...[
        round(
          mean([
            convertPressure(bytes[2], bytes[3]),
            convertPressure(bytes[5], bytes[6]),
            convertPressure(bytes[8], bytes[9]),
          ]),
          2,
        ),
        round(
          mean([
            convertPressure(bytes[11], bytes[12]),
            convertPressure(bytes[14], bytes[15]),
            convertPressure(bytes[17], bytes[18]),
          ]),
          2,
        ),
      ],
    );
    temps.push(
      ...[
        round(
          mean([
            convertSensorTemperature(bytes[4]),
            convertSensorTemperature(bytes[7]),
            convertSensorTemperature(bytes[10]),
          ]),
          2,
        ),
        round(
          mean([
            convertSensorTemperature(bytes[13]),
            convertSensorTemperature(bytes[16]),
            convertSensorTemperature(bytes[19]),
          ]),
          2,
        ),
      ],
    );
  }
  return {pressures, temperatures: temps};
};
