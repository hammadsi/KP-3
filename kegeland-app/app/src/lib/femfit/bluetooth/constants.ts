import {BluetoothService} from '~constants/bluetooth/interface';
import {transformId} from '~utils/platform.utils';

export const PRESSURE_SENSOR_MIN_VAL = 760;
export const PRESSURE_SENSOR_MAX_VAL = 970;
export const ACTIVATION_THRESHOLD = 0.15;

/**
 * Transform UUIDs based on the current platform
 */
const HOUSEKEEPING_SERVICE_UUID = transformId(
  'b92a0001-4bf9-4870-8aa1-881b3a20ada4',
);
const HOUSEKEEPING_CHARACTERISTIC_UUID = transformId(
  'b92a0002-4bf9-4870-8aa1-881b3a20ada4',
);
const SENSOR_SERVICE_UUID = transformId('0d9e0001-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_1_UUID = transformId('0d9e0002-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_2_UUID = transformId('0d9e0003-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_3_UUID = transformId('0d9e0004-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_4_UUID = transformId('0d9e0005-c111-49cd-bba3-85c7471cb6fa');

/**
 * The femfit services
 */
export const SERVICES = {
  HOUSEKEEPING_SERVICE: HOUSEKEEPING_SERVICE_UUID,
  SENSOR_SERVICE: SENSOR_SERVICE_UUID,
};

/**
 * The housekeeping service for femfit
 */
export const HOUSEKEEPING_SERVICE: BluetoothService = {
  uuid: HOUSEKEEPING_SERVICE_UUID,
  characteristics: [HOUSEKEEPING_CHARACTERISTIC_UUID],
};

/**
 * The sensor service for femfit
 */
export const SENSOR_SERVICE: BluetoothService = {
  uuid: SENSOR_SERVICE_UUID,
  characteristics: [SENSOR_1_UUID, SENSOR_2_UUID, SENSOR_3_UUID, SENSOR_4_UUID],
};
