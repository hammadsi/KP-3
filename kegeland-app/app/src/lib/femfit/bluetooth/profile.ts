import {BluetoothProfile} from '~constants/bluetooth/interface';

import {HOUSEKEEPING_SERVICE, SENSOR_SERVICE} from './constants';
import {readBattery} from './utils';

/**
 * The bluetooth profile for femfit
 */
export default {
  navScreen: 'Femfit',
  services: [HOUSEKEEPING_SERVICE.uuid, SENSOR_SERVICE.uuid],
  batteryService: {
    service: HOUSEKEEPING_SERVICE,
    fn: readBattery,
  },
} as BluetoothProfile;
