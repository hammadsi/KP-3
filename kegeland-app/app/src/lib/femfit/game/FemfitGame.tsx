import React, {ReactNode, PureComponent} from 'react';
import {StatusBar, StyleSheet, View, Image} from 'react-native';
import SystemNavigationBar from 'react-native-system-navigation-bar';
import {GameEngine} from 'react-native-game-engine';
import {max, reduce, round, set} from 'lodash';
import {connect, ConnectedProps} from 'react-redux';

import BackgroundImage from '~assets/femfit/background.png';
import {BluetoothDevice} from '~state/ducks/bluetooth/bluetooth.interface';
import {PeripheralNotification} from '~constants/bluetooth/interface';
import {addServiceListener, removeServiceListener} from '~utils/bluetooth';
import {bleManagerEmitter} from '~hooks/useBluetooth';
import {UPDATE_INTERVAL_MS} from '~constants/bluetooth';
import {DeviceScreenProps} from '~routes/interface';
import {setSession} from '~state/ducks/session/session.reducer';
import {clearAnswers} from '~state/ducks/questions/questions.reducer';

import {ACTIVATION_THRESHOLD, SENSOR_SERVICE} from '../bluetooth/constants';
import {pressurePercent, readSensorBytes} from '../bluetooth/utils';

import GameOverDialog from './components/GameOverDialog';
import MenuDialog from './components/MenuDialog';
import Header from './components/Header';
import spawnEntities from './entities/spawnEntities';

import {ExerciseScheme, GameEntities} from './interface';
import Physics from './physics';
import constants from './constants';

const {MAX_HEIGHT, MAX_WIDTH} = constants;

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  setSession,
  clearAnswers,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

export type FemfitGameProps = {
  device: BluetoothDevice;
  exercise: ExerciseScheme;
  navigation: DeviceScreenProps<'Femfit'>['navigation'];
} & PropsFromRedux;

type FemfitGameState = {
  running: boolean;
  gameOver: boolean;
  score: number;
  gameOverDialogVisible: boolean;
  sessionData: Record<number, number[]>;
};

/**
 * Component for rendering the gamified exercise for femfit devices
 */
class FemfitGame extends PureComponent<FemfitGameProps, FemfitGameState> {
  gameEngine: any | null;
  exerciseScheme: ExerciseScheme;
  entities: GameEntities;
  sensorData: Record<string, number[]>;
  interval: any;
  constructor(props: any) {
    super(props);
    this.gameEngine = null;
    this.exerciseScheme = this.props.exercise;
    // Initialize the game entities
    this.entities = spawnEntities(this.exerciseScheme);
    // Initialize empty sensor data
    this.sensorData = reduce(
      SENSOR_SERVICE.characteristics,
      (prev, curr) => {
        prev[curr] = [];
        return prev;
      },
      {} as Record<string, number[]>,
    );
    this.interval = null;
    // Initialize the state
    this.state = {
      running: true,
      gameOver: false,
      score: 0,
      gameOverDialogVisible: false,
      sessionData: {},
    };
  }

  /**
   * Function for opening/closing the menu
   */
  toggleMenuDialog() {
    this.setState({...this.state, running: !this.state.running});
  }

  /**
   * Function for opening/closing the game over dialog
   */
  toggleGameOverDialog() {
    this.setState({
      ...this.state,
      gameOverDialogVisible: !this.state.gameOverDialogVisible,
    });
  }

  /**
   * Handles notification updates from the femfit sensor
   * @param param0 the update data
   * @see {@link PeripheralNotification}
   */
  handlePeripheralUpdate({characteristic, value}: PeripheralNotification) {
    if (characteristic in this.sensorData) {
      set(this.sensorData, characteristic, value);
    }
  }

  /**
   * Logic executed when component is mounted
   */
  componentDidMount(): void {
    // Hide the navigation bar
    SystemNavigationBar.navigationHide();
    this.props.navigation.setOptions({headerShown: false});
    // Reset the entities
    this.gameEngine.swap(spawnEntities(this.exerciseScheme));
    this.gameEngine.dispatch({type: 'reset'});
    // Initialize listeners for the sensor
    if (this.props.device && this.props.device.state === 'connected') {
      addServiceListener(this.props.device.id, SENSOR_SERVICE);
      bleManagerEmitter.addListener(
        'BleManagerDidUpdateValueForCharacteristic',
        (data: PeripheralNotification) => this.handlePeripheralUpdate(data),
      );
      // Start interval for reading the sensor data
      this.interval = setInterval(() => {
        const {pressures, temperatures} = readSensorBytes(
          Object.values(this.sensorData),
        );
        const ts = Date.now();
        // Update the state with new session data
        this.setState({
          ...this.state,
          sessionData: {
            ...this.state.sessionData,
            [ts]: [...pressures, ...temperatures],
          },
        });
        const pctMax = round(pressurePercent(max(pressures) as number), 1);
        // Move the character if the pressure is above threshold
        if (pctMax > ACTIVATION_THRESHOLD) {
          this.gameEngine.dispatch({type: 'move_up', value: pctMax});
        }
      }, UPDATE_INTERVAL_MS);
    }
  }

  /**
   * Logic executed when the component dismounts
   */
  componentWillUnmount(): void {
    // Show the navigation bar
    SystemNavigationBar.navigationShow();
    this.props.navigation.setOptions({headerShown: true});
    // Clear the update interval
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
    // Unsubscribe the bluetooth listeners
    removeServiceListener(this.props.device.id, SENSOR_SERVICE);
    bleManagerEmitter.removeAllListeners(
      'BleManagerDidUpdateValueForCharacteristic',
    );
  }

  render(): ReactNode {
    return (
      <View style={styles.container}>
        <Image
          source={BackgroundImage}
          style={styles.backgroundImage}
          resizeMode="stretch"
        />
        <Header
          score={this.state.score}
          toggle={() => this.toggleMenuDialog()}
        />
        <View style={styles.container}>
          <StatusBar hidden />
          <GameEngine
            ref={(ref) => {
              this.gameEngine = ref;
            }}
            running={this.state.running}
            entities={this.entities}
            style={styles.gameContainer}
            onEvent={(event: any) => {
              switch (event.type) {
                case 'score':
                  this.setState({
                    ...this.state,
                    score: this.state.score + 1,
                  });
                  break;
                case 'game_over':
                  if (!this.state.gameOver) {
                    this.setState(
                      {
                        ...this.state,
                        running: false,
                        gameOver: true,
                      },
                      () => {
                        this.props.setSession({
                          sensor: 'femfit',
                          data: this.state.sessionData,
                        });
                        this.toggleGameOverDialog();
                      },
                    );
                  }
                  break;
              }
            }}
            systems={[Physics]}
          />
        </View>
        <MenuDialog
          visible={!this.state.running && !this.state.gameOver}
          onDismiss={() => this.toggleMenuDialog()}
          goBack={() => {
            this.props.clearAnswers();
            this.props.navigation.goBack();
          }}
        />
        <GameOverDialog
          score={this.state.score}
          exercise={this.exerciseScheme}
          visible={this.state.gameOverDialogVisible}
          onDismiss={() => this.toggleGameOverDialog()}
          goBack={() => {
            this.props.navigation.goBack();
          }}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    zIndex: 999,
  },
  backgroundImage: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    width: MAX_WIDTH,
    height: MAX_HEIGHT,
  },
  gameContainer: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
  },
});

export default connector(FemfitGame);
