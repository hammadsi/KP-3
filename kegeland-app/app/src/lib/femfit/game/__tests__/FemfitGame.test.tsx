import {cloneDeep, set} from 'lodash';
import React from 'react';

import {HOUSEKEEPING_SERVICE} from '~lib/femfit/bluetooth/constants';
import {BluetoothState} from '~state/ducks/bluetooth/bluetooth.interface';
import {initialState} from '~state/ducks/bluetooth/bluetooth.reducer';
import {deviceMock} from '~state/ducks/__mocks__/bluetooth.mocks';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';
import {render} from '~utils/test-utils';

import FemfitGame, {FemfitGameProps} from '../FemfitGame';
import exerciseMock from '../__mocks__/exercise.mock';
jest.useFakeTimers();
describe('Test FemfitGame-component', () => {
  const housekeepingService = HOUSEKEEPING_SERVICE;
  const bluetoothState: BluetoothState = {
    ...initialState,
    connectedDevices: {[deviceMock.id]: deviceMock},
    housekeepers: {[housekeepingService.uuid]: deviceMock.id},
  };

  const store = mockStore(
    set(cloneDeep(initialStore), 'bluetooth', bluetoothState),
  );
  it('should render correctly', () => {
    const navigation: any = {
      navigate: jest.fn(),
      setOptions: jest.fn(),
    };

    const props: FemfitGameProps = {
      device: deviceMock,
      setSession: jest.fn() as any,
      exercise: exerciseMock,
      clearAnswers: jest.fn() as any,
      navigation,
    };
    const component = <FemfitGame {...props} />;
    const tree = render(component, {wrapperProps: {store}});
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
