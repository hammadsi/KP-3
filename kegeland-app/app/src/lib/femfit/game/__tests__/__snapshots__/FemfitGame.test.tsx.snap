// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FemfitGame-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "zIndex": 999,
      }
    }
  >
    <Image
      resizeMode="stretch"
      source={
        Object {
          "testUri": "../../../assets/femfit/background.png",
        }
      }
      style={
        Object {
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 8,
            "zIndex": 1000,
          },
          Object {
            "backgroundColor": "#ffffff",
            "height": 50,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "fontSize": 20,
                "fontWeight": "bold",
              },
            ]
          }
        >
          0
        </Text>
        <Image
          resizeMode="stretch"
          source={
            Object {
              "testUri": "../../../assets/femfit/gold.png",
            }
          }
          style={
            Object {
              "height": 25,
              "marginLeft": 8,
              "width": 25,
            }
          }
        />
      </View>
      <View>
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={1}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            nativeID="animatedComponent"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#00b0ff",
                    "fontSize": 28,
                  },
                  Array [
                    Object {
                      "alignSelf": "center",
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              Û∞çú
            </Text>
          </View>
        </RNGestureHandlerButton>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "zIndex": 999,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            }
          }
        />
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
